src/test/regress/upgrade/config.py

```diff
@@ -1,68 +1,290 @@
 
-class PGUpgradeConfig():
+    def setup_steps(self):
+        pass
+
+    def random_worker_port(self):
+        return random.choice(self.worker_ports)
+
+    def _init_node_name_ports(self):
+        self.node_name_to_ports = {}
+        self.worker_ports = []
+        cur_port = self._get_and_update_next_port()
+        self.node_name_to_ports[COORDINATOR_NAME] = cur_port
+        for i in range(self.worker_amount):
+            cur_port = self._get_and_update_next_port()
+            cur_worker_name = "worker{}".format(i)
+            self.node_name_to_ports[cur_worker_name] = cur_port
+            self.worker_ports.append(cur_port)
+
+    def _get_and_update_next_port(self):
+        if hasattr(self, "fixed_port"):
+            next_port = self.fixed_port
+            self.fixed_port += 1
+            return next_port
+        return find_free_port()
+
+
+class CitusMXBaseClusterConfig(CitusBaseClusterConfig):
     def __init__(self, arguments):
-        self.old_bindir = arguments['--old-bindir']
-        self.new_bindir = arguments['--new-bindir']
-        self.pg_srcdir = arguments['--pgxsdir']
-        self.temp_dir = './tmp_upgrade'
-        self.old_datadir = self.temp_dir + '/oldData'
-        self.new_datadir = self.temp_dir + '/newData'
-        self.settings = {
-            'shared_preload_libraries': 'citus',
-            'citus.node_conninfo': 'sslmode=prefer'
+        super().__init__(arguments)
+        self.is_mx = True
+        self.add_coordinator_to_metadata = True
+
+
+class CitusUpgradeConfig(CitusBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.pre_tar_path = arguments["--citus-pre-tar"]
+        self.post_tar_path = arguments["--citus-post-tar"]
+        self.temp_dir = "./tmp_citus_upgrade"
+        self.new_settings = {"citus.enable_version_checks": "false"}
+        self.user = SUPER_USER_NAME
+        self.mixed_mode = arguments["--mixed"]
+        self.fixed_port = 57635
+
+
+class CitusDefaultClusterConfig(CitusBaseClusterConfig):
+    pass
+
+
+class PostgresConfig(CitusDefaultClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.worker_amount = 0
+        self.is_citus = False
+
+
+class CitusSingleNodeClusterConfig(CitusDefaultClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.worker_amount = 0
+
+
+class CitusSingleWorkerClusterConfig(CitusMXBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.worker_amount = 1
+
+
+class CitusSuperUserDefaultClusterConfig(CitusMXBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.user = SUPER_USER_NAME
+
+
+class CitusFiveWorkersManyShardsClusterConfig(CitusMXBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.new_settings = {"citus.shard_count": 494}
```

Comment:
This is causing a lot of pressure on my local machine, and some queries fail -- especially repartition joins.

Shall we make it like, 191 shards or such?

Reply:
Sounds good